From 7f97eef40420826cbb7f769a2d6d94fea2edd33a Mon Sep 17 00:00:00 2001
From: Pavel Sobolev <pavel@paveloom.dev>
Date: Tue, 23 Jul 2024 22:19:17 +0300
Subject: [PATCH] Add CORS headers to the http responses (#418)

---
 src/ios_webkit_debug_proxy.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/src/ios_webkit_debug_proxy.c b/src/ios_webkit_debug_proxy.c
index fe066aa4..284b353e 100644
--- a/src/ios_webkit_debug_proxy.c
+++ b/src/ios_webkit_debug_proxy.c
@@ -679,7 +679,9 @@ ws_status iwdp_send_http(ws_t ws, bool is_head, const char *status,
   if (asprintf(&data,
       "HTTP/1.1 %s\r\n"
       "Content-length: %zd\r\n"
-      "Connection: close"
+      "Connection: close\r\n"
+      "Access-Control-Allow-Origin: *\r\n"
+      "Access-Control-Allow-Methods: GET, HEAD"
       "%s%s\r\n\r\n%s",
       status, (content ? strlen(content) : 0),
       (ctype ? "\r\nContent-Type: " : ""), (ctype ? ctype : ""),
