diff --git a/bin/phase-helpers.sh b/bin/phase-helpers.sh
index aa9f6af8b..5c83ea498 100644
--- a/bin/phase-helpers.sh
+++ b/bin/phase-helpers.sh
@@ -1021,29 +1021,52 @@ if ___eapi_has_eapply; then
 	}
 
 	eapply() {
-		local LC_ALL LC_COLLATE=C f i path
+		local LC_ALL LC_COLLATE=C seen_doubledash path i f
 		local -a operands options
 
+		# PMS mandates an unconventional option parsing scheme whereby
+		# the rule that options must precede non-option arguments is
+		# only enforced in the case that no "--" argument is found.
+		#
+		# https://projects.gentoo.org/pms/8/pms.html#x1-127001r1
 		while (( $# )); do
 			case $1 in
 				--)
+					seen_doubledash=1
 					shift
-					operands+=("$@")
 					break
 					;;
-				-*)
-					if (( ! ${#operands[@]} )); then
-						options+=("$1")
-					else
-						die "eapply: options must precede non-option arguments"
-					fi
-					;;
 				*)
-					operands+=("$1")
+					options+=("$1")
 			esac
 			shift
 		done
 
+		if (( seen_doubledash )); then
+			# The "--" argument was encountered. Forward any to its
+			# its left to the patch(1) utility, verbatim.
+			operands=("$@")
+		else
+			# Restore the positional parameters and parse normally.
+			set -- "${options[@]}" "$@"
+			options=()
+
+			while (( $# )); do
+				case $1 in
+					-*)
+						if (( ! ${#operands[@]} )); then
+							options+=("$1")
+						else
+							die "eapply: options must precede non-option arguments"
+						fi
+						;;
+					*)
+						operands+=("$1")
+				esac
+			shift
+			done
+		fi
+
 		if (( ! ${#operands[@]} )); then
 			die "eapply: no operands were specified"
 		fi
